Common/AR/source/Android/ArCoreWrapper.cpp

```diff
@@ -0,0 +1,933 @@
+    return {};
+  }
+
+  ArCamera_getImageIntrinsics(m_arSession, m_arCamera, cameraIntrinsics);
+
+  float xFocalLength = 0.0f;
+  float yFocalLength = 0.0f;
+  ArCameraIntrinsics_getFocalLength(m_arSession, cameraIntrinsics, &xFocalLength, &yFocalLength);
+
+  float xPrincipal = 0.0f;
+  float yPrincipal = 0.0f;
+  ArCameraIntrinsics_getPrincipalPoint(m_arSession, cameraIntrinsics, &xPrincipal, &yPrincipal);
+
+  int32_t xImageSize = 0;
+  int32_t yImageSize = 0;
+  ArCameraIntrinsics_getImageDimensions(m_arSession, cameraIntrinsics, &xImageSize, &yImageSize);
+
+  ArCameraIntrinsics_destroy(cameraIntrinsics);
+
+  return
+  {
+    static_cast<double>(xFocalLength), static_cast<double>(yFocalLength),
+    static_cast<double>(xPrincipal), static_cast<double>(yPrincipal),
+    static_cast<double>(xImageSize), static_cast<double>(yImageSize)
+  };
+}
+
+/*!
+  \internal
+  Implementation based on the documentation: https://developers.google.com/ar/reference/c/group/frame#arframe_hittest
+ */
+std::array<double, 7> ArCoreWrapper::hitTest(int x, int y) const
+{
+  if (!m_arSession || !m_arCamera)
+    return {};
+
+  // create hit result list
+  ArHitResultList* hitResults = nullptr;
+  ArHitResultList_create(m_arSession, &hitResults);
+  if (!hitResults)
+    return {};
+
+  // try to find the location point
+  ArFrame_hitTest(m_arSession, m_arFrame, x, y, hitResults);
+
+  int32_t size = 0;
+  ArHitResultList_getSize(m_arSession, hitResults, &size);
+  if (size <= 0)
+  {
+    ArHitResultList_destroy(hitResults);
+    return {};
+  }
+
+  ArHitResult* hitResult = nullptr;
+  ArHitResult_create(m_arSession, &hitResult);
+  if (!hitResult)
+  {
+    ArHitResultList_destroy(hitResults);
+    return {};
+  }
+
+  ArHitResultList_getItem(m_arSession, hitResults, 0, hitResult);
+
+  ArPose* pose = nullptr;
+  ArPose_create(m_arSession, nullptr, &pose);
+  if (!hitResult)
+  {
+    ArHitResult_destroy(hitResult);
+    ArHitResultList_destroy(hitResults);
+    return {};
+  }
+
+  ArHitResult_getHitPose(m_arSession, hitResult, pose);
+
+  float poseRaw[7] = {};
+  ArPose_getPoseRaw(m_arSession, pose, poseRaw);
+
+  // destroys objects
+  ArPose_destroy(pose);
+  ArHitResult_destroy(hitResult);
+  ArHitResultList_destroy(hitResults);
+
+  return { poseRaw[0], poseRaw[1], poseRaw[2], poseRaw[3], poseRaw[4], poseRaw[5], poseRaw[6] };
+}
+
+/*!
+  \internal
+ */
+std::array<float, 8> ArCoreWrapper::transformedUvs() const
+{
+  return m_transformedUvs;
+}
+
+/*!
+  \internal
+  properties for debug mode
+ */
+QColor ArCoreWrapper::pointCloudColor() const
+{
+  if (m_arCorePointCloudRenderer)
+    return m_arCorePointCloudRenderer->pointCloudColor();
+  return QColor();
+}
+
+/*!
+  \internal
+ */
+void ArCoreWrapper::setPointCloudColor(const QColor& pointCloudColor)
+{
+  if (pointCloudColor.isValid())
+  {
+    if (!m_arCorePointCloudRenderer)
+    {
+      m_arCorePointCloudRenderer.reset(new ArCorePointCloudRenderer(this));
```

Comment:
Same question here about ownership. I won't comment on all of these. We need to figure out if we're doing QObject parent ownership or std::unique_ptr ownership; we shouldn't do both unless there is a reason to.

Reply:
Same. All the `XxxRenderer` classes are not `QObject`, they need to access to the wrapper to get the data.

It's a little bit tricky, because on Android, the renderer is called by the wrapper (and then the data are sent when the function is called) and it's not necessary to keep a pointer to the wrapper in the renderer. And on iOS, the renderer is called by the `QQuickItem` and the renderer calls the wrapper functions (and then the wrapper need to keep the data until the renderer use it).
I think it's ok for the concurrent access. In `ArKitWrapper`, the property `textureDataUsed` is used to avoid to delete the frame texture if the `ArKitFrameRenderer` still use the texture. This is the reason I'm using `std::vector` in parameter (to copy the data) rather than using `float*`.